v4/export/dump.go

```diff
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
+	}
+	return min, max, nil
+}
+
+func (d *Dumper) concurrentDumpTiDBTables(conn *sql.Conn, meta TableMeta, ir chan<- TableDataIR) error {
+	ctx, conf := d.ctx, d.conf
+	db, tbl := meta.DatabaseName(), meta.TableName()
+
+	handleColNames, handleVals, err := selectTiDBTableSample(db, tbl, conn, meta)
+	if err != nil {
+		return err
+	}
+	if len(handleVals) == 0 {
+		return nil
+	}
+	selectField, selectLen, err := buildSelectField(conn, db, tbl, conf.CompleteInsert)
+	if err != nil {
+		return err
+	}
+	where := buildWhereClauses(handleColNames, handleVals)
+	orderByClause := buildOrderByClauseString(handleColNames)
+
+	for _, w := range where {
+		query := buildSelectQuery(db, tbl, selectField, buildWhereCondition(conf, w), orderByClause)
+		ctxDone := sendDataIRToChan(ctx, newTableData(query, selectLen), ir)
+		if ctxDone {
+			break
+		}
+	}
+	return nil
+}
+
+func selectTiDBTableSample(dbName, tableName string, db *sql.Conn, meta TableMeta) (handleCols []string, pkVals []string, err error) {
+	pkFields, err := GetPrimaryKeyColumns(db, dbName, tableName)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	query := buildTiDBTableSampleQuery(pkFields, dbName, tableName)
+	handleCols, handleColTypes := findHandleColsAndColTypes(pkFields, meta)
+	rows, err := db.QueryContext(context.Background(), query)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	iter := newRowIter(rows, len(handleCols))
+	defer iter.Close()
+	rowRec := MakeRowReceiver(handleColTypes)
+	buf := new(bytes.Buffer)
+	for iter.HasNext() {
+		err = iter.Decode(rowRec)
 		if err != nil {
-			return nil, err
+			return nil, nil, errors.Trace(err)
 		}
-		if err := writer.WriteTableMeta(ctx, dbName, tableName, createTableSQL); err != nil {
-			return nil, err
+		rowRec.WriteToBuffer(buf, true)
+		pkVals = append(pkVals, buf.String())
+		buf.Reset()
+		iter.Next()
+	}
+	return handleCols, pkVals, nil
+}
+
+func buildTiDBTableSampleQuery(pkFields []string, dbName, tblName string) string {
+	hasImplicitRowID := pkFields == nil
+	if hasImplicitRowID {
+		template := "SELECT `_tidb_rowid` FROM `%s`.`%s` TABLESAMPLE REGIONS() ORDER BY `_tidb_rowid`"
+		return fmt.Sprintf(template, dbName, tblName)
+	}
+	template := "SELECT %s FROM `%s`.`%s` TABLESAMPLE REGIONS() ORDER BY %s"
+	quotaPk := make([]string, len(pkFields))
+	for i, s := range pkFields {
+		quotaPk[i] = fmt.Sprintf("`%s`", s)
+	}
+	pks := strings.Join(quotaPk, ",")
+	return fmt.Sprintf(template, pks, dbName, tblName, pks)
+}
+
+func findHandleColsAndColTypes(pkFields []string, meta TableMeta) (handleCols []string, colTypes []string) {
+	hasImplicitRowID := pkFields == nil
+	if hasImplicitRowID {
+		return []string{"_tidb_rowid"}, []string{"BIGINT"}
+	}
+	// Find primary key column types in meta.
+	colTypes = make([]string, len(pkFields))
+	for i, pkColName := range pkFields {
+		for j, colName := range meta.ColumnNames() {
+			if pkColName == colName {
+				colTypes[i] = meta.ColumnTypes()[j]
```

Comment:
We will build a string array every time we use `meta.ColumnTypes`. It looks a little strange.
What about saving `columnNames` and `columnTypes` instead of `[]*sql.ColumnType` at first?

Reply:
What's more, I think we can select both primary key columns and their types through one SQL from MySQL/TiDB. I think that's a better implementation than this way.